### PR TITLE
Update Template and TemplateCategory models

### PR DESCRIPTION
**Pull request by whabanks**
# Summary | Résumé

- `Template.category` is now `Template.template_category`
- In `Template` and `TemplateHistory`, the `template_category` attribute now leverage a primary join to: 
    1. Allow the full TemplateCategory model to be loaded when fetching Template or TemplateHistory 
    2. Prevent a full list of associated templates from being populated on a TemplateCategory model
- Updated the Template and TemplateHistory schemas to include the TemplateCategory when returning data to API callers

## Related Issues | Cartes liées

# Test instructions | Instructions pour tester la modification

_TODO: Fill in test instructions for the reviewer._

# Release Instructions | Instructions pour le déploiement

None.

# Reviewer checklist | Liste de vérification du réviseur

- [ ] This PR does not break existing functionality.
- [ ] This PR does not violate GCNotify's privacy policies.
- [ ] This PR does not raise new security concerns. Refer to our GC Notify Risk Register document on our Google drive.
- [ ] This PR does not significantly alter performance.
- [ ] Additional required documentation resulting of these changes is covered (such as the README, setup instructions, a related ADR or the technical documentation).

> ⚠ If boxes cannot be checked off before merging the PR, they should be moved to the "Release Instructions" section with appropriate steps required to verify before release. For example, changes to celery code may require tests on staging to verify that performance has not been affected.

### CODE REVIEW

File: app/dao/template_categories_dao.py [hunk 1]
```diff
@@ -69,12 +68,12 @@ def dao_delete_template_category_by_id(template_category_id, cascade=False):
...
                     else template_category.email_process_type
                 )
-                template.category = dao_get_template_category_by_id(default_category_id)
+                template.template_category_id = dao_get_template_category_by_id(default_category_id).id
                 template.updated_at = datetime.utcnow()
                 db.session.add(template)
+                db.session.commit()
 
         db.session.delete(template_category)
-        db.session.commit()
```
Comment (sastels) [inline app/dao/template_categories_dao.py:77]: Why did you move the commit?

File: app/models.py [hunk 2]
```diff
@@ -1117,8 +1117,8 @@ def template_category_id(cls):
         return db.Column(UUID(as_uuid=True), db.ForeignKey("template_categories.id"), index=True, nullable=True)
 
     @declared_attr
-    def category(cls):
-        return db.relationship("TemplateCategory")
+    def template_category(cls):
+        return db.relationship("TemplateCategory", primaryjoin="Template.template_category_id == TemplateCategory.id")
```
Comment (jzbahrai) [inline app/models.py:1121]: nice

File: app/schemas.py [hunk 3]
```diff
@@ -396,6 +396,7 @@ class TemplateSchema(BaseTemplateSchema):
     created_by = field_for(models.Template, "created_by", required=True)
     is_precompiled_letter = fields.Method("get_is_precompiled_letter")
     process_type = field_for(models.Template, "process_type")
+    template_category = fields.Nested(TemplateCategorySchema, dump_only=True)
```
Comment (jzbahrai) [inline app/schemas.py:399]: ah I missed adding the sms_sending_vehicle can we add it in this PR

File: app/dao/template_categories_dao.py [hunk 4]
```diff
@@ -69,12 +68,12 @@ def dao_delete_template_category_by_id(template_category_id, cascade=False):
                     if template.template_type == "sms"
                     else template_category.email_process_type
                 )
-                template.category = dao_get_template_category_by_id(default_category_id)
+                template.template_category_id = dao_get_template_category_by_id(default_category_id).id
```
[the remark this answers is not in the document]
Comment (whabanks) [reply app/dao/template_categories_dao.py:71]: Yea we definitely don't.. Remnant from when we were getting and setting the model itself rather than the id. I'll update it.

File: app/dao/template_categories_dao.py (hunk 1 above)
Comment (whabanks) [reply app/dao/template_categories_dao.py:77]: I removed it because with `@transactional` we auto commit after the method executes. Buuuut in testing I had removed `@transactional`. I'll add it back in.

File: app/schemas.py (hunk 3 above)
Comment (whabanks) [reply app/schemas.py:399]: It's a simple type and the `TemplateCategorySchema` references our model. Since you updated it there it picks it up okay in the front end. Here's what it looks like locally in Redis:

```
{
  "id": "d59f9609-b965-4836-bb51-904ee9f3c2b7",
  "name_en": "hmm",
  "name_fr": "hmm",
  "description_en": "hmm",
  "description_fr": "hmm",
  "sms_process_type": "normal",
  "email_process_type": "priority",
  "hidden": false,
  "created_at": "2024-07-08T19:05:13.129712",
  "updated_at": null,
  "sms_sending_vehicle": "long_code"
}
```
template_categories:
```
{
    "id": "b6c42a7e-2a26-4a07-802b-123a5c3198a9",
    "name_en": "Authentication",
    "name_fr": "Authentification",
    "description_en": "Password resets and two factor verification",
    "description_fr": "R\u00e9initialisations de mots de passe et v\u00e9rification \u00e0 deux facteurs",
    "sms_process_type": "priority",
    "email_process_type": "priority",
    "hidden": false,
    "created_at": "2024-07-08T17:03:08.501236",
    "updated_at": "2024-07-08T17:03:08.501236",
    "sms_sending_vehicle": "short_code"
  },
  {
    "id": "977e2a00-f957-4ff0-92f2-ca3286b24786",
    "name_en": "Automatic reply",
    "name_fr": "R\u00e9ponse automatique",
    "description_en": "No-reply and confirmation messages",
    "description_fr": "Messages auxquels il est impossible de r\u00e9pondre et messages de confirmation",
    "sms_process_type": "priority",
    "email_process_type": "priority",
    "hidden": false,
    "created_at": "2024-07-08T17:03:08.501236",
    "updated_at": "2024-07-08T17:03:08.501236",
    "sms_sending_vehicle": "short_code"
  },
....
```

**Review by jzbahrai — APPROVED** [review]
